test/distributed/org/apache/cassandra/distributed/upgrade/ConsistencyTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.upgrade;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.Test;
+
+import org.apache.cassandra.distributed.UpgradeableCluster;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
+import org.apache.cassandra.distributed.api.Feature;
+import org.apache.cassandra.distributed.api.ICoordinator;
+import org.apache.cassandra.distributed.api.IUpgradeableInstance;
+import org.apache.cassandra.distributed.shared.Versions;
+
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.*;
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
+
+public class ConsistencyTest extends UpgradeTestBase
+{
+    @Test
+    public void testConsistency() throws Throwable
+    {
+        testConsistency(allUpgrades(3, 1));
+    }
+
+    @Test
+    public void testConsistencyWithNetworkAndGossip() throws Throwable
+    {
+        testConsistency(new TestCase().nodes(3)
+                                      .nodesToUpgrade(1)
+                                      // .upgrade(Versions.Major.v22, Versions.Major.v30)
+                                      // .upgrade(Versions.Major.v22, Versions.Major.v3X)
+                                      // .upgrade(Versions.Major.v30, Versions.Major.v3X)
```

Comment:
@maedhroz I don't know why using these versions make the test fail with an `UnavailableException` on reads.

Reply:
@adelapena I ran through the 3.0 -> 3.11 upgrade example. Once node 1 is upgraded, attempting a read from node 2 as a coordinator at ALL fails thusly: `Live nodes [/127.0.0.2, /127.0.0.3] do not satisfy ConsistencyLevel (3 required)`. Why doesn't node 2 think node 1 has rejoined the cluster?

Reply:
Yeah, the reason this works without `NETWORK/GOSSIP` is that the endpoint data is manually maintained there. I'm not 100% sure, but I don't know if we really need to test this with GOSSIP enabled at all. We're testing coordination behavior through upgrades, assuming gossip and other networking systems are functioning normally.

@adelapena WDYT about just avoiding NETWORK and GOSSIP in this patch?

Reply:
There may also be a way to work around it here: https://github.com/apache/cassandra/blob/cassandra-3.11/test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeReadTest.java#L71-L78

Thanks to @jrwest for the link.

Reply:
Unfortunately those methods are gone in trunk. Probably we don't really need gossip on these tests, so I have left the original `UpgradeTest#simpleUpgradeWithNetworkAndGossipTest` test as it was.